Index: root/projects/3rd-party/distro/yui-2.9.0-patched/META-INF/yui/datatable/datatable-debug.js
===================================================================
--- root/projects/3rd-party/distro/yui-2.9.0-patched/META-INF/yui/datatable/datatable-debug.js	(revision 6)
+++ root/projects/3rd-party/distro/yui-2.9.0-patched/META-INF/yui/datatable/datatable-debug.js	(working copy)
@@ -5185,6 +5185,7 @@
                     var elOldTh = oOldColumn.getThEl();
                     Dom.removeClass(elOldTh, oOldSortedBy.dir);
                     this.formatTheadCell(oOldColumn.getThLinerEl().firstChild, oOldColumn, oNewSortedBy);
+                    oOldColumn.getThEl().setAttribute("aria-sort","none");
                 }
                 if(oNewSortedBy) {
                     oNewColumn = (oNewSortedBy.column) ? oNewSortedBy.column : this._oColumnSet.getColumn(oNewSortedBy.key);
@@ -5204,6 +5205,14 @@
                          Dom.addClass(elNewTh, sortClass);
                     }
                     this.formatTheadCell(oNewColumn.getThLinerEl().firstChild, oNewColumn, oNewSortedBy);
+                    if (oNewSortedBy.dir == DT.CLASS_ASC)
+                    {
+                       oNewColumn.getThEl().setAttribute("aria-sort","ascending");
+                    }
+                    else
+                    {
+                       oNewColumn.getThEl().setAttribute("aria-sort","descending");
+                    }
                 }
             }
           
@@ -6527,6 +6536,10 @@
     }
 
     this.formatTheadCell(elThLabel, oColumn, this.get("sortedBy"));
+    if (oColumn.sortable)
+    {
+       oColumn.getThEl().setAttribute("aria-sort","none");
+    }
     oColumn._elThLabel = elThLabel;
 },
 
