Index: core/src/test/java/net/sf/ehcache/jcache/JCacheTest.java
===================================================================
--- core/src/test/java/net/sf/ehcache/jcache/JCacheTest.java	(revision 516)
+++ core/src/test/java/net/sf/ehcache/jcache/JCacheTest.java	(working copy)
@@ -16,9 +16,9 @@
 
 package net.sf.ehcache.jcache;
 
-import edu.emory.mathcs.backport.java.util.concurrent.ExecutionException;
-import edu.emory.mathcs.backport.java.util.concurrent.ExecutorService;
-import edu.emory.mathcs.backport.java.util.concurrent.Future;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
 import net.sf.ehcache.AbstractCacheTest;
 import net.sf.ehcache.Ehcache;
 import net.sf.ehcache.StopWatch;
Index: core/src/main/java/net/sf/ehcache/distribution/MulticastKeepaliveHeartbeatReceiver.java
===================================================================
--- core/src/main/java/net/sf/ehcache/distribution/MulticastKeepaliveHeartbeatReceiver.java	(revision 516)
+++ core/src/main/java/net/sf/ehcache/distribution/MulticastKeepaliveHeartbeatReceiver.java	(working copy)
@@ -31,8 +31,8 @@
 import java.util.StringTokenizer;
 import java.util.List;
 
-import edu.emory.mathcs.backport.java.util.concurrent.ExecutorService;
-import edu.emory.mathcs.backport.java.util.concurrent.Executors;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 
 /**
  * Receives heartbeats from any {@link MulticastKeepaliveHeartbeatSender}s out there.
Index: core/src/main/java/net/sf/ehcache/constructs/asynchronous/AsynchronousCommandExecutor.java
===================================================================
--- core/src/main/java/net/sf/ehcache/constructs/asynchronous/AsynchronousCommandExecutor.java	(revision 516)
+++ core/src/main/java/net/sf/ehcache/constructs/asynchronous/AsynchronousCommandExecutor.java	(working copy)
@@ -28,8 +28,8 @@
 import java.util.Date;
 import java.util.Stack;
 
-import edu.emory.mathcs.backport.java.util.Queue;
-import edu.emory.mathcs.backport.java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
 
 /**
  * Handles the asynchronous execution of commands. This class contains subtle threading interactions and should
Index: core/src/main/java/net/sf/ehcache/jcache/JCache.java
===================================================================
--- core/src/main/java/net/sf/ehcache/jcache/JCache.java	(revision 516)
+++ core/src/main/java/net/sf/ehcache/jcache/JCache.java	(working copy)
@@ -16,11 +16,11 @@
 
 package net.sf.ehcache.jcache;
 
-import edu.emory.mathcs.backport.java.util.concurrent.ExecutionException;
-import edu.emory.mathcs.backport.java.util.concurrent.Future;
-import edu.emory.mathcs.backport.java.util.concurrent.LinkedBlockingQueue;
-import edu.emory.mathcs.backport.java.util.concurrent.ThreadPoolExecutor;
-import edu.emory.mathcs.backport.java.util.concurrent.TimeUnit;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
 import net.sf.ehcache.Ehcache;
 import net.sf.ehcache.Element;
 import net.sf.jsr107cache.CacheEntry;
